include/gz/math/TimeVaryingVolumetricGridLookupField.hh

```diff
@@ -0,0 +1,217 @@
+
+      /// \brief Adds a volumetric grid field.
+      public: void AddVolumetricGridField(
+        const T& time, const VolumetricGridLookupField<V> &_field) {}
+
+      /// \brief Creates a session for querying
+      public: S CreateSession() {}
+
+      public: S CreateSession(const T &_time) {}
+
+      public: std::optional<S> StepTo(const S &_session, const T &_time) {}
+
+      public: std::pair<InterpolationPoint4D<T, V>, InterpolationPoint4D<T, V>>
+      LookUp(const S &_session,
+        const T &_time,
+        const Vector3<V> &_point,
+        const Vector3<V> &_tol) {}
+
+      public: template<typename X>
+      X EstimateQuadrilinear(
+        const S &_session,
+        const std::pair<InterpolationPoint4D<T, V>,
+          InterpolationPoint4D<T, V>> &_points,
+        const std::vector<X>& values1,
+        const std::vector<X>& values2
+      )
+      {
+      }
+    };
+    /// \brief An in-memory session. Loads the whole dataset in memory and
+    /// performs queries.
+    template<typename T, typename V>
+    struct InMemorySession
+    {
+      /// \brief Iterator which holds pointer to current state
+      /// TODO(arjo): Use friend to make visible only to InMemorySession
+      /// specialization
+      typename std::map<T, VolumetricGridLookupField<V>>::iterator iter;
+    };
+
+    /// \brief Specialized version of `TimeVaryingVolumetricGridLookupField`
+    /// for in-memory lookup. It loads the whole dataset into memory.
+    template<typename T, typename V>
+    class TimeVaryingVolumetricGridLookupField<T, V, InMemorySession<T, V>>
+    {
+      public: TimeVaryingVolumetricGridLookupField()
+      {}
+
+      public: void AddVolumetricGridField(
+        const T &_time, const VolumetricGridLookupField<V> &_field) {
+        this->gridFields.emplace(_time, _field);
+      }
+
+      public: InMemorySession<T, V> CreateSession() {
+        InMemorySession<T, V> sess;
+        sess.iter = this->gridFields.begin();
+        return sess;
+      }
+
+      public: InMemorySession<T, V> CreateSession(const T &_time) {
+        InMemorySession<T, V> sess;
+        sess.iter = this->gridFields.lower_bound(_time);
+        return sess;
+      }
+
+      public: std::optional<InMemorySession<T, V>> StepTo(
+        const InMemorySession<T, V> &_session, const T &_time) {
+        if (_session.iter == gridFields.end())
+        {
+          return std::nullopt;
+        }
+
+        InMemorySession<T, V> newSess(_session);
+
+        auto nextTime = std::next(_session.iter);
+        if (nextTime == this->gridFields.end() || _time < _session.iter->first)
+        {
+          return std::nullopt;
+        }
+
+        while (nextTime != this->gridFields.end()
+          && nextTime->first <= _time)
+        {
+          newSess.iter = nextTime;
+          nextTime = std::next(nextTime);
+        }
+        return newSess;
+      }
+
+      public: std::pair<InterpolationPoint4D<T, V>, InterpolationPoint4D<T, V>>
```

Comment:
Fair enough we could just add another `AxisIndex` to `VolumetricGridFieldData`. This does have slight performance advantage for this usecase though.

Reply:
> This does have slight performance advantage for this usecase though.

In what sense?

Reply:
By keeping track of the session iter we can use `std::next` which has amortized O(1) (although worst case O(logn)) IIRC. That being said I agree I can simplify this by just reusing AxisIndex and save us this headache.

Reply:
Right, but N-dimensional arrays will outperform hash maps and binary trees if accessed in a cache conscious way.

Reply:
So I had originally proposed N-d arrays to Ben and mabel however, given precision conversion errors between UTM<=> and Lat/Lon, the grids are not quite consistent. Ben was the one who actually told u to use trees as thats what works for mbari. W.r.t to time we could probably get away assuming our data is well ordered and use an `std::vector` instead of std::map then. Arbitrary lookup would still rely on a binary search using `std::lower_bound`

Reply:
@arjo129 ahh, I see now. Having a session/cursor for the index makes total sense.